File: bubble_sort/python/bubble_sort.py
import sys
import time

def bubble_sort(nums):
    """
    Bubble Sort O(n^2)
    Implementação in-place com otimização de parada antecipada.
    """
    n = len(nums)
    for i in range(n - 1):
        swapped = False
        for j in range(n - i - 1):
            if nums[j] > nums[j + 1]:
                nums[j], nums[j + 1] = nums[j + 1], nums[j]
                swapped = True
        if not swapped:
            break  # Melhor caso: O(n)

def rodar_teste(nome_arquivo, n):
    try:
        # Lê o arquivo e converte para lista de floats
        with open(nome_arquivo, 'r') as f:
            # Carrega apenas as primeiras 'n' linhas
            v = []
            for _ in range(n):
                line = f.readline()
                if not line:
                    break
                v.append(float(line.strip()))

        if len(v) < n:
            print(f"Aviso: O arquivo {nome_arquivo} possui apenas {len(v)} registros.")

        # Medição de tempo usando time.perf_counter() (alta precisão no Linux)
        start = time.perf_counter()
        bubble_sort(v)
        end = time.perf_counter()

        return end - start
    except FileNotFoundError:
        print(f"Erro: Arquivo {nome_arquivo} não encontrado.")
        return -1
    except Exception as e:
        print(f"Erro: {e}")
        return -1

def main():
    if len(sys.argv) < 2:
        print(f"Uso: python3 {sys.argv[0]} <n_elementos>")
        sys.exit(1)

    n = int(sys.argv[1])
    print(f"--- Iniciando Benchmark Python (N: {n}) ---\n")

    testes = [
        ("Crescente", "crescente.txt", "(Melhor caso)"),
        ("Aleatorio", "aleatorio.txt", "(Caso Médio)"),
        ("Decrescente", "decrescente.txt", "(Pior caso)")
    ]

    for nome, path, desc in testes:
        tempo = rodar_teste(path, n)
        if tempo >= 0:
            print(f"{testes.index((nome, path, desc)) + 1}. {nome:12}: {tempo:.6f} segundos {desc}")

    print("\n--- Testes concluidos ---")

if __name__ == "__main__":
    main()

File: bubble_sort/javascript/bubble_sort.js
const fs = require('fs');
const { performance } = require('perf_hooks');

/**
 * Bubble Sort O(n^2)
 * Implementação in-place com otimização de parada antecipada.
 */
function bubbleSort(nums) {
    let n = nums.length;
    for (let i = 0; i < n - 1; i++) {
        let swapped = false;
        for (let j = 0; j < n - i - 1; j++) {
            if (nums[j] > nums[j + 1]) {
                let temp = nums[j];
                nums[j] = nums[j + 1];
                nums[j + 1] = temp;
                swapped = true;
            }
        }
        if (!swapped) break; // Melhor caso: O(n)
    }
}

/**
 * Função para carregar dados e medir o tempo.
 */
function rodarTeste(nomeArquivo, n) {
    try {
        // Lê o arquivo de forma síncrona para simplificar o benchmark
        const data = fs.readFileSync(nomeArquivo, 'utf8');
        // Converte as linhas em números e limita a 'n' elementos
        let v = data.split('\n')
                    .filter(line => line.trim() !== '')
                    .slice(0, n)
                    .map(Number);

        if (v.length < n) {
            console.log(`Aviso: O arquivo ${nomeArquivo} possui apenas ${v.length} registros.`);
        }

        const start = performance.now();
        bubbleSort(v);
        const end = performance.now();

        // performance.now() retorna milissegundos, convertemos para segundos
        return (end - start) / 1000;
    } catch (err) {
        console.error(`Erro ao abrir ${nomeArquivo}:`, err.message);
        return -1;
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log("Uso: node bubble_sort.js <n_elementos>");
        process.exit(1);
    }

    const n = parseInt(args[0]);
    console.log(`--- Iniciando Benchmark JavaScript (N: ${n}) ---\n`);

    const testes = [
        { nome: "Crescente", path: "crescente.txt", desc: "(Melhor caso)" },
        { nome: "Aleatorio", path: "aleatorio.txt", desc: "(Caso Médio)" },
        { nome: "Decrescente", path: "decrescente.txt", desc: "(Pior caso)" }
    ];

    testes.forEach((teste, index) => {
        const tempo = rodarTeste(teste.path, n);
        if (tempo >= 0) {
            console.log(`${index + 1}. ${teste.nome.padEnd(12)}: ${tempo.toFixed(6)} segundos ${teste.desc}`);
        }
    });

    console.log("\n--- Testes concluidos ---");
}

main();